### ISSUE TITLE
Add Karpenter support

### ISSUE THREAD

**Issue by Skarlso** [kubernetes-sigs/cluster-api#7198]
**User Story**

I would like to use Karpenter as auto-scaler.

**Detailed Description**

Karpenter provides a nice way of configuring autoscaling options. Right now, there is only support for AWS but more can / will be added later on with providers mechanism.

CAPI can support Karpenter being installed as an autoscaler.

https://karpenter.sh/

**Anything else you would like to add:**

Some talk has already been initiated about Karpetner supporting installation into CAPI/CAPA clusters here: https://github.com/aws/karpenter-core/issues/747

And another one was opened in CAPA, but was decided/suggested that this should be implemented in CAPI since its like the cluster-autoscaler here: https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/3564

/kind feature

**Comment by sbueringer** [comment]
cc @randomvariable @elmiko (just fyi)

**Comment by fabriziopandini** [comment]
/triage needs-information

while I'm generally +1 to support tools in the ecosystem, I think we should qualify a little bit better what are the implication in cluster-API for this request; TBD if this requires a proposal

**Comment by Skarlso** [comment]
👍 For sure. I'm not very well versed with CAPI so I'm unsure what the implications are and what they look like. I require some help in determining that. :)

**Comment by fabriziopandini** [comment]
that's fair. I think we can start collecting what karpenter needs, then someone can step in to help in defining how this can be implemented in CAPI

**Comment by Skarlso** [comment]
Awesome. Sounds good. I can add what little information I know. :) The attached Karpenter ticket details some of the things that Karpenter itself is working on to make this a viable thing. It might even be that we have to wait for them a bit to come up with something.

**Comment by elmiko** [comment]
thanks for the ping @sbueringer , definitely sounds like a good use case to me. from my understanding of karpenter i imagine there would be work to be done around synchronizing the instances created by it with the records that cluster-api keeps. perhaps this might be best used with MachinePools?

as i understand it, karpenter wants to talk directly to the AWS infrastructure. i had looked into a cluster-api provider for karpenter but had trouble integrating with their efforts at the time, i would imagine it's stabilized some since then. i have a feeling that deploying karpenter in a cluster-api cluster will look slightly different than with the cluster-autoscaler mainly due to the differences with how they both talk to the infrastructure.

with all that said, i have a feeling that deploying karpenter inside an aws would work, but i'm not sure what effect it would have on the MachineDeployments/Sets as those would get out of sync with its decision making. additionally, i believe karpenter wants to have access to a variety of instance sizes for it to create, this is where linking to a MachinePool style resource might make more sense.

happy to see this brought up =)

**Comment by Skarlso** [comment]
/remove-lifecycle rotten
Close this issue with /close

**Comment by elmiko** [comment]
we talked about this issue at the community meeting, i am going to see if there is interest in forming a feature group to investigate capi/karpenter integration. ideally, i will put a poll up in our slack channel sometime next week to see if we have enough interest to meet up.

**Comment by elmiko** [comment]
just to update here, the karpenter feature group has been meeting for several months now and we have agreed on a design and are working towards a proof-of-concept to demonstrate for the wider cluster-api community.

**Comment by fabriziopandini** [comment]
@elmiko It would be great to share the design with the broader community to get a first quick round of feedback

**Comment by elmiko** [comment]
@fabriziopandini absolutely. i think it's taken some time to land on what will work for us, i am preparing a slidedeck to present at an upcoming capi meeting to help further the discussion.

**Comment by dtzar** [comment]
For reference, there is a feature group meeting for this: https://hackmd.io/@elmiko/ryR2VXR0n

**Comment by memorais** [comment]
Hi there, as cluster admin I'm interested to see this working. If any help is needed here, I would be happy to participate.

**Comment by elmiko** [comment]
@dntosas yes it is!

we have contributed the initial version to the cncf, you can find it here https://github.com/kubernetes-sigs/karpenter-provider-cluster-api

and we have feature group meeting every other wednesday, see more here https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/community/20231018-karpenter-integration.md

perhaps we should close this issue, but i don't think we have fully landed on our final architecture. that said, the project works and we are now iterating on different architecture designs and addressing bugs.

**Comment by Skarlso** [comment]
Oh that is absolutely fantastic. :)

**Comment by elmiko** [comment]
perhaps we should update the docs, but unless there is a further request here i think we should close this issue.

https://github.com/kubernetes-sigs/karpenter-provider-cluster-api has just released version 0.2.0 and is generally stable for testing by cluster-api users. it is still considered in an alpha state, but we welcome all contributions. =)

**Comment by fabriziopandini** [comment]
Thanks very much @elmiko to drive this!
I encourage everyone to take a look at https://github.com/kubernetes-sigs/karpenter-provider-cluster-api, test it, provide feedback!     

/close
